control-plane/connect-inject/peering_acceptor_controller.go

```diff
@@ -0,0 +1,259 @@
+
+	// TODO(peering): If the status is empty, then what should be done? This means a peering exists in Consul but hasn't
+	// been reconciled by the controller, or there was an error during a previous reconciliation. I think this means we
+	// should continue and generate a new token and update the status with that.
+	_, existingSecret, err := r.getExistingSecret(ctx, peeringAcceptor.Status.Secret.Name, peeringAcceptor.Namespace)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	shouldGenerate, err := r.shouldGenerateToken(peeringAcceptor, existingSecret)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if shouldGenerate {
+		// Generate and store the peering token.
+		var resp *api.PeeringGenerateTokenResponse
+		if resp, err = r.generateToken(ctx, peeringAcceptor.Name); err != nil {
+			return ctrl.Result{}, err
+		}
+		if peeringAcceptor.Spec.Peer.Secret.Backend == "kubernetes" {
+			if err := r.createK8sPeeringTokenSecretWithOwner(ctx, peeringAcceptor, resp); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		// Store the state in the status.
+		err := r.updateStatus(ctx, peeringAcceptor, resp)
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *PeeringAcceptorController) shouldGenerateToken(peeringAcceptor *consulv1alpha1.PeeringAcceptor, existingSecret *corev1.Secret) (bool, error) {
+	if peeringAcceptor.Status.Secret == nil || peeringAcceptor.Status.LastReconcileTime == nil {
+		return false, errors.New("shouldGenerateToken was called with an empty fields in the existing status")
+	}
+	// Compare the existing name, key, and backend.
+	if peeringAcceptor.Status.Secret.Name != peeringAcceptor.Spec.Peer.Secret.Name {
+		return true, nil
+	}
+	if peeringAcceptor.Status.Secret.Key != peeringAcceptor.Spec.Peer.Secret.Key {
+		return true, nil
+	}
+	// TODO(peering): remove this when validation webhook exists.
+	if peeringAcceptor.Status.Secret.Backend != peeringAcceptor.Spec.Peer.Secret.Backend {
+		return false, errors.New("PeeringAcceptor backend cannot be changed")
+	}
+	// Compare the existing secret hash.
+	// Get the secret specified by the status, make sure it matches the status' secret.latestHash.
+	if existingSecret != nil {
+		existingSecretHashBytes := sha256.Sum256(existingSecret.Data[peeringAcceptor.Status.Secret.Key])
+		existingSecretHash := hex.EncodeToString(existingSecretHashBytes[:])
+		if existingSecretHash != peeringAcceptor.Status.Secret.LatestHash {
+			r.Log.Info("secret doesn't match status.secret.latestHash, should generate new token")
+			return true, nil
+		}
+
+	}
+	return false, nil
+}
+func (r *PeeringAcceptorController) updateStatus(ctx context.Context, peeringAcceptor *consulv1alpha1.PeeringAcceptor, resp *api.PeeringGenerateTokenResponse) error {
+	peeringAcceptor.Status.Secret = &consulv1alpha1.SecretStatus{
+		Name:    peeringAcceptor.Spec.Peer.Secret.Name,
+		Key:     peeringAcceptor.Spec.Peer.Secret.Key,
+		Backend: peeringAcceptor.Spec.Peer.Secret.Backend,
```

Comment:
some of these can be simplified by adding methods to the PeeringAcceptor like SecretName() that will return spec.peer.secret.name so that it is more readable and you get the benefits of adding Nil checks.

Reply:
This is a great idea, going to do a passthrough to do this.

Reply:
update: haven't gotten to this yet

Reply:
I ended up just adding an Asana to clean this up later in favor of merging sooner: https://app.asana.com/0/1201832349181086/1202342432709154

Reply:
perfect! thanks Nitya!